From bc837067663c4c7e46824ea9695a64beb60dae28 Mon Sep 17 00:00:00 2001
From: Sidhanta Sahu <quic_sidhanta@quicinc.com>
Date: Mon, 25 Mar 2024 11:53:02 -0700
Subject: [PATCH] hostapd: Fix mld link dl list crash

Currently, the hapd->link is added to the mld links dl list during
driver init and setup BSS operation. However, the below call trace is
seen when an BSS link item is not present in the dl list and tries to
delete it from the list. This scenario occurs during the interface
deinit operation which calls bss link deinit and tries to remove the
hapd which is not present in the dl list.

The fix is to make sure the link item is added to the dl list after
successful operation of link addition and make sure the mld->num_links
increments only when it is successful. So during bss link deinit
returns if mld->num_links is zero. Since the mld object is shared among
all the links, so the num_links needs to be incremented only when
successful.

Call trace:
	0x00000055560c9610 in dl_list_del.lto_priv.9.lto_priv ()
	0x00000055560ca944 in hostapd_bss_link_deinit.lto_priv ()
	0x00000055560d3680 in hostapd_bss_deinit ()
	0x00000055560d5d90 in hostapd_interface_deinit ()
	0x00000055560d5e14 in hostapd_interface_deinit_free ()
	0x00000055560c5e40 in hostapd_main ()

Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 hostapd/main.c   | 12 ++++++++----
 src/ap/hostapd.c | 13 +++++++++----
 src/ap/ubus.c    |  8 ++++----
 3 files changed, 21 insertions(+), 12 deletions(-)

--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -198,7 +198,6 @@ static int hostapd_driver_init(struct ho
 			os_memcpy(hapd->own_addr, b, ETH_ALEN);
 		}
 
-		hostapd_mld_add_link(hapd);
 		wpa_printf(MSG_DEBUG, "Setup of non first link BSS(link id %d) of MLD %s",
 			   hapd->mld_link_id, hapd->conf->iface);
 
@@ -282,7 +281,6 @@ pre_setup_mld:
 		else
 			os_memcpy(hapd->own_addr, b, ETH_ALEN);
 
-		hostapd_mld_add_link(hapd);
 		wpa_printf(MSG_DEBUG, "Setup of first link (link %d) BSS of MLD %s",
 			   hapd->mld_link_id, hapd->conf->iface);
 	}
@@ -340,8 +338,14 @@ setup_mld:
 			   hapd->mld_link_id, MAC2STR(hapd->mld->mld_addr),
 			   MAC2STR(hapd->own_addr));
 
-		hostapd_drv_link_add(hapd, hapd->mld_link_id,
-				     hapd->own_addr);
+		if (hostapd_drv_link_add(hapd, hapd->mld_link_id,
+					 hapd->own_addr)) {
+			wpa_printf(MSG_ERROR,
+				   "MLD: Failed to add link %d in MLD %s",
+				   hapd->mld_link_id, hapd->conf->iface);
+			return -1;
+		}
+		hostapd_mld_add_link(hapd);
 	}
 #endif /* CONFIG_IEEE80211BE */
 
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1451,7 +1451,6 @@ static int hostapd_setup_bss(struct host
 			if (h_hapd) {
 				hapd->drv_priv = h_hapd->drv_priv;
 				hapd->interface_added = h_hapd->interface_added;
-				hostapd_mld_add_link(hapd);
 				wpa_printf(MSG_DEBUG, "TSetup of non first link BSS(link id %d) of MLD %s",
 					   hapd->mld_link_id, hapd->conf->iface);
 				goto setup_mld;
@@ -1477,7 +1476,6 @@ static int hostapd_setup_bss(struct host
 			wpa_printf(MSG_DEBUG, "TSetup of first link BSS of MLD %s",
 				   hapd->conf->iface);
 			os_memcpy(hapd->mld->mld_addr, hapd->own_addr, ETH_ALEN);
-			hostapd_mld_add_link(hapd);
 		}
 	}
 
@@ -1707,8 +1705,14 @@ setup_mld:
 			   hapd->mld_link_id, MAC2STR(hapd->mld->mld_addr),
 			   MAC2STR(hapd->own_addr));
 
-		hostapd_drv_link_add(hapd, hapd->mld_link_id,
-				     hapd->own_addr);
+		if (hostapd_drv_link_add(hapd, hapd->mld_link_id,
+					 hapd->own_addr)) {
+			wpa_printf(MSG_ERROR,
+				   "MLD: Failed to add link %d in MLD %s",
+				   hapd->mld_link_id, hapd->conf->iface);
+			return -1;
+		}
+		hostapd_mld_add_link(hapd);
 	}
 #endif /* CONFIG_IEEE80211BE */
 
--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -1730,6 +1730,10 @@ void hostapd_ubus_add_bss(struct hostapd
 	obj->methods = bss_object_type.methods;
 	obj->n_methods = bss_object_type.n_methods;
 	ret = ubus_add_object(ctx, obj);
+	if (ret) {
+		free(name);
+		obj->name = NULL;
+	}
 	hostapd_ubus_ref_inc();
 
 	hostapd_send_shared_event(&hapd->iface->interfaces->ubus, hapd->conf->iface, "add");
@@ -1752,9 +1756,14 @@ void hostapd_ubus_free_bss(struct hostap
 
 	if (obj->id)
 		ubus_remove_object(ctx, obj);
+
 	hostapd_ubus_ref_dec();
 
+	if (!name)
+		return;
+
 	free(name);
+	obj->name = NULL;
 }
 
 static void
